scripts/test-new-tests.mjs

```diff
@@ -0,0 +1,143 @@
+
+  const remoteUrl =
+    eventData?.head?.repo?.full_name ||
+    process.env.GITHUB_REPOSITORY ||
+    (await execa('git remote get-url origin', EXECA_OPTS)).stdout
+
+  const isCanary =
+    branchName.trim() === 'canary' && remoteUrl.includes('vercel/next.js')
+
+  if (isCanary) {
+    console.error(`Skipping flake detection for canary`)
+    return
+  }
+
+  try {
+    await execa('git remote set-branches --add origin canary', EXECA_OPTS_STDIO)
+    await execa('git fetch origin canary --depth=20', EXECA_OPTS_STDIO)
+  } catch (err) {
+    console.error(await execa('git remote -v', EXECA_OPTS_STDIO))
+    console.error(`Failed to fetch origin/canary`, err)
+  }
+
+  const changesResult = await execa(
+    `git diff origin/canary --name-only`,
+    EXECA_OPTS
+  ).catch((err) => {
+    console.error(err)
+    return { stdout: '', stderr: '' }
+  })
+  console.error(
+    {
+      branchName,
+      remoteUrl,
+      isCanary,
+    },
+    `\ngit diff:\n${changesResult.stderr}\n${changesResult.stdout}`
+  )
+  const changedFiles = changesResult.stdout.split('\n')
+
+  // run each test 3 times in each test mode (if E2E) with no-retrying
+  // and if any fail it's flakey
+  const devTests = []
+  const prodTests = []
+
+  for (let file of changedFiles) {
+    // normalize slashes
+    file = file.replace(/\\/g, '/')
+    const fileExists = await fs
+      .access(path.join(process.cwd(), file), fs.constants.F_OK)
+      .then(() => true)
+      .catch(() => false)
+
+    if (fileExists && file.match(/^test\/.*?\.test\.(js|ts|tsx)$/)) {
+      if (file.startsWith('test/e2e/')) {
+        devTests.push(file)
+        prodTests.push(file)
+      } else if (file.startsWith('test/prod')) {
+        devTests.push(file)
+      } else if (file.startsWith('test/development')) {
+        prodTests.push(file)
+      }
+    }
+  }
+
+  console.log(
+    'Detected tests:',
+    JSON.stringify(
+      {
+        devTests,
+        prodTests,
+      },
+      null,
+      2
+    )
+  )
+
+  if (prodTests.length === 0 && devTests.length === 0) {
+    console.log(`No added/changed tests detected`)
+    return
+  }
+
+  const RUN_TESTS_ARGS = ['run-tests.js', '-c', '1', '--retries', '0']
+
+  async function invokeRunTests({ mode, testFiles }) {
+    await execa('node', [...RUN_TESTS_ARGS, ...devTests], {
+      ...EXECA_OPTS_STDIO,
+      env: {
+        ...process.env,
+        NEXT_TEST_MODE: mode,
+      },
+    })
+  }
+
+  if (devTests.length > 0) {
+    for (let i = 0; i < 3; i++) {
+      console.log(`\n\nRun ${i + 1} for dev tests`)
+      await invokeRunTests({
+        mode: 'dev',
+        testFiles: devTests,
+      })
+    }
+  }
+
+  if (prodTests.length > 0) {
+    for (let i = 0; i < 3; i++) {
+      console.log(`\n\nRun ${i + 1} for production tests`)
+      await invokeRunTests({
+        mode: 'start',
+        testFiles: prodTests,
+      })
+    }
+  }
+}
+
+main().catch((err) => {
+  console.error(err)
+  process.exit(1)
+})
```

Comment:
Do you need a then here to not exit the process before `main()` concludes? Or better yet, is there any output that we want to log then this completes successfully?

```suggestion
main().then(() => { process.exit(0) }).catch((err) => {
  console.error(err)
  process.exit(1)
})
```

Reply:
It should exit cleanly on it's own, if we need `process.exit(0)` something is unexpectedly hanging